data_structures/raster_source.cpp

```diff
@@ -0,0 +1,183 @@
+}
+
+// Query raster source for nearest data point
+RasterDatum RasterSource::getRasterData(const float lon, const float lat) const
+{
+    int _lon = static_cast<int>(lon * COORDINATE_PRECISION);
+    int _lat = static_cast<int>(lat * COORDINATE_PRECISION);
+
+    if (_lon < xmin || _lon > xmax || _lat < ymin || _lat > ymax)
+    {
+        return {};
+    }
+
+    float xthP = (_lon - xmin) / xstep;
+    int xth = ((xthP - floor(xthP)) > (xstep / 2.) ? floor(xthP) : ceil(xthP));
+
+    float ythP = (ymax - _lat) / ystep;
+    int yth = ((ythP - floor(ythP)) > (ystep / 2.) ? floor(ythP) : ceil(ythP));
+
+    return {raster_data(xth, yth)};
+}
+
+// Query raster source using bilinear interpolation
+RasterDatum RasterSource::getRasterInterpolate(const float lon, const float lat) const
+{
+    int _lon = static_cast<int>(lon * COORDINATE_PRECISION);
+    int _lat = static_cast<int>(lat * COORDINATE_PRECISION);
+
+    if (_lon < xmin || _lon > xmax || _lat < ymin || _lat > ymax)
+    {
+        return {};
+    }
+
+    unsigned xthP = (_lon - xmin) / xstep;
+    unsigned ythP = (ymax - _lat) / ystep;
+    int top = floor(ythP);
+    int bottom = ceil(ythP);
+    int left = floor(xthP);
+    int right = ceil(xthP);
+
+    float x = (_lon - left * xstep + xmin) / xstep;
+    float y = (ymax - top * ystep - _lat) / ystep;
+    float x1 = 1.0 - x;
+    float y1 = 1.0 - y;
+
+    return {static_cast<std::int32_t>(
+        raster_data(left, top) * (x1 * y1) + raster_data(right, top) * (x * y1) +
+        raster_data(left, bottom) * (x1 * y) + raster_data(right, bottom) * (x * y))};
+}
+
+// Load raster source into memory
+int SourceContainer::loadRasterSource(const std::string &source_path,
+                                      double xmin,
+                                      double xmax,
+                                      double ymin,
+                                      double ymax,
+                                      unsigned nrows,
+                                      unsigned ncols)
+{
+    int _xmin = static_cast<int>(xmin * COORDINATE_PRECISION);
+    int _xmax = static_cast<int>(xmax * COORDINATE_PRECISION);
+    int _ymin = static_cast<int>(ymin * COORDINATE_PRECISION);
+    int _ymax = static_cast<int>(ymax * COORDINATE_PRECISION);
+
+    auto itr = LoadedSourcePaths.find(source_path);
+    if (itr != LoadedSourcePaths.end())
+    {
+        std::cout << "[source loader] Already loaded source '" << source_path << "' at source_id "
+                  << itr->second << std::endl;
+        return itr->second;
+    }
+
+    int source_id = LoadedSources.size();
+
+    std::cout << "[source loader] Loading from " << source_path << "  ... " << std::flush;
+    TIMER_START(loading_source);
+
+    if (!boost::filesystem::exists(source_path.c_str()))
+    {
+        throw osrm::exception("error reading: no such path");
+    }
+
+    RasterGrid rasterData{source_path, ncols, nrows};
+
+    RasterSource source{rasterData,
```

Comment:
You do not need the `rasterData` variable from here on, as you pass it to the `source` variable's constructor (sink). Therefore, instead of 1/ a copy operation into the `source` constructor and then 2/ the `rasterData` goes out of scope and gets destructed anyway just do a single move:

    RasterSource source{std::move(rasterData), ..